### PR TITLE
Refactor: Replace scan diff hook options by new commands

### PR DESCRIPTION
**Pull request by Paul-GitGuardian**
Currently, the `iac scan diff` command handles three options to simplify git hooks: `--pre-commit`, `--pre-push` and `--pre-receive`. The objective of the `main-iac-cli-hooks` integration branch is to replace them with new `iac scan` subcommands.

This PR will refactor elements of the code in order to:
- make the implementation of the subcommands (not in the scope of this PR) easier
- factorize code

The duplicate code in `precommit.py`, `prepush.py` and `prereceive.py` can be ignored for now.

### CODE REVIEW

**Comment by GG-HH** [comment]
Sounds good to me except the files that you asked not to review.

File: ggshield/cmd/iac/scan/all.py [hunk 1]
```diff
@@ -74,3 +73,13 @@ def iac_scan_all(ctx: click.Context, directory: Path) -> Optional[IaCScanResult]
         handle_scan_error(client, scan)
         return None
     return scan
+
+
+def display_iac_scan_all_result(
+    ctx: click.Context,
+    directory: Path,
+    result: Union[IaCScanResult, None],
```
Comment (agateau-gg) [inline ggshield/cmd/iac/scan/all.py:81]: nit: should be
```suggestion
    result: Optional[IaCScanResult],
```

File: ggshield/cmd/iac/scan/iac_scan_common_options.py [hunk 2]
```diff
@@ -65,10 +65,18 @@ def _validate_exclude(_ctx: Any, _param: Any, value: Sequence[str]) -> Sequence[
     help="Do not scan the specified paths.",
 )
 
+all_option = click.option(
+    "--all",
+    is_flag=True,
+    default=None,
+    help="Raise all vulnerabilities in the final state.",
```
Comment (agateau-gg) [inline ggshield/cmd/iac/scan/iac_scan_common_options.py:72]: wording:
```suggestion
    help="Report all vulnerabilities in the final state.",
```

File: ggshield/cmd/iac/scan/precommit.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    **kwargs: Any,
+) -> int:
+    """
+    Scan as pre-commit for IaC vulnerabilities. By default, it will return vulnerabilities added in the commit.
+    """
+    display_warning(
+        "This feature is still in beta, its behavior may change in future versions."
+    )
+
+    # TODO: WIP
```
Comment (agateau-gg) [inline ggshield/cmd/iac/scan/precommit.py:40]: This comment can be removed: it does not tell readers what to do here.

File: ggshield/cmd/iac/scan/precommit.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    if directory is None:
+        directory = Path().resolve()
+    update_context(ctx, exit_zero, minimum_severity, ignore_policies, ignore_paths)
+
+    # TODO: remove this once the GGClient is updated with the new diff function
```
Comment (agateau-gg) [inline ggshield/cmd/iac/scan/precommit.py:46]: I think this comment is no longer necessary?

File: ggshield/cmd/iac/scan/precommit.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    config: Config = ctx.obj["config"]
+    ctx.obj["client"] = create_client_from_config(config)
+
+    ref = "HEAD"
+    staged = True
```
Comment (agateau-gg) [inline ggshield/cmd/iac/scan/precommit.py:51]: maybe inline those?

(this comment and the two previous ones apply to the other `scan_pre_*_cmd()` functions as well)

**Review by agateau-gg — CHANGES_REQUESTED** [review]
Minor remarks, looks good overall. Do you need the `main-iac-cli-hooks` branch though? It seems to me these changes could be merged into `main` directly.

**Comment by Paul-GitGuardian** [comment]
I agree regarding the branch, I expected it to be needed but it isn't. I've rebased on main directly.
Unfortunately, I can't edit the target branch in the PR, so I've created a new one (#619) and will close this one once the other is merged on main.

**Pull request closed without merging**